[MLIR][Vector] Added ToElementsOp::fold for broadcast->to_elements pattern rewrite.

keshavvinayak01 opened the pull request:
Adds `::fold` for the new `vector.to_elements` op, folding `broadcast` into `to_elements` or no-op wherever possible.

keshavvinayak01 committed:
Added folder for broadcast->to_elements

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

Groverkss commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2395,11 +2395,103 @@ foldToElementsFromElements(ToElementsOp toElementsOp,
+
+  // Create a temporary to_elements to get the source elements for mapping.
+  // Change the operand to the broadcast source.
+  OpBuilder builder(toElementsOp);
+  auto srcElems = builder.create<ToElementsOp>(toElementsOp.getLoc(),
+                                               bcastOp.getSource());
```
You can't do this in a folder. Look at the documentation for reference: https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-the-fold-method . If you have to create a Rewriter/Builder, something is wrong. You can just modify the operation in place. Have a look at some other examples in VectorOps.cpp for reference.

keshavvinayak01 replied:
As per our discussion, we need a canonicalizer for it.

keshavvinayak01 committed:
Vector broadcast->to_elements requires a canonicalizer

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Removed duplicate lit

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

Groverkss commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2441,88 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+    auto loc = toElementsOp.getLoc();
+    auto srcElems = rewriter.create<ToElementsOp>(loc, bcastOp.getSource());
+
+    int64_t dstCount = 1;
+    for (int64_t v : dstShape)
+      dstCount *= v;
```
Can use std::accumulate here

Groverkss commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2441,88 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+    if (srcRank > dstRank)
+      return failure();
+
+    // Verify broadcastability (right-aligned)
```
What is a right aligned broadcast?

keshavvinayak01 replied:
This check is pointless tbh, I don't know why I added it earlier. Thanks for being thorough.

Groverkss commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2441,88 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+      return failure();
+
+    // Verify broadcastability (right-aligned)
+    for (unsigned i = 0; i < dstRank; ++i) {
+      int64_t dstDim = dstShape[i];
+      int64_t srcDim = 1;
+      if (i + srcRank >= dstRank)
+        srcDim = srcShape[i + srcRank - dstRank];
+      if (!(srcDim == 1 || srcDim == dstDim))
+        return failure();
+    }
```
I'm not sure what you are trying to verify here, can you write an example of what this is

Groverkss commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2441,88 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
 
+namespace {
+
+struct ToElementsOfVectorBroadcast final
```
Add doc with an example of what this does

Groverkss commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2441,88 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+
+    // Pre-compute and apply mapping from destination linear index to
+    // source linear index (row-major, right-aligned broadcasting).
+    SmallVector<int64_t> dstIdx(dstShape.size());
+    for (int64_t lin = 0; lin < dstCount; ++lin) {
+      int64_t temp = lin;
+      for (int64_t i = dstShape.size() - 1; i >= 0; --i) {
+        int64_t dim = dstShape[i];
+        dstIdx[i] = temp % dim;
+        temp /= dim;
+      }
+      int64_t srcLin = 0;
+      for (unsigned k = 0; k < srcRank; ++k)
+        srcLin = srcLin * srcShape[k] +
+                 ((srcShape[k] == 1) ? 0 : dstIdx[dstRank - srcRank + k]);
+
+      replacements.push_back(srcElems.getResult(srcLin));
+    }
+
```
This is very hard to understand, please write more docs and add an example. Also, please don't use new terminology like "row-major, right-aligned".

keshavvinayak01 replied:
You're right. I simplified the comments

keshavvinayak01 committed:
> Addressed review comments:
1. Better comments, removing keywords and verbose docs.
2. Removed redundant "Broadcastability" check, we don't require it since the vector.broadcast op will always be valid when it reaches this logic.

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2441,88 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+
+    auto dstType = cast<VectorType>(toElementsOp.getSource().getType());
+
+    // Bail on scalable vectors.
```
Done

keshavvinayak01 commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2441,88 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+    }
+
+    // Create elements for the broadcast source vector.
+    auto loc = toElementsOp.getLoc();
```
Done

keshavvinayak01 committed:
Added doc for the canonicalizer

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2443,91 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+    SmallVector<Value> replacements;
+    replacements.reserve(dstCount);
+
+    // For each element of the destination, determine which element of the
+    // source should be used. We walk all destination positions using a single
+    // counter, decode it into per-dimension indices, then build the matching
+    // source position: use the same index where sizes match, and use 0 where
+    // the source size is 1 (replication). This mapping is needed so we can
+    // replace each result of to_elements with the corresponding element from
+    // the broadcast source.
+    SmallVector<int64_t> dstIdx(dstShape.size());
+    for (int64_t lin = 0; lin < dstCount; ++lin) {
+      int64_t temp = lin;
+      for (int64_t i = dstShape.size() - 1; i >= 0; --i) {
+        int64_t dim = dstShape[i];
+        dstIdx[i] = temp % dim;
+        temp /= dim;
+      }
+      int64_t srcLin = 0;
+      for (int64_t k = 0; k < srcRank; ++k)
+        srcLin = srcLin * srcShape[k] +
+                 ((srcShape[k] == 1) ? 0 : dstIdx[dstRank - srcRank + k]);
+
+      replacements.push_back(srcElems.getResult(srcLin));
+    }
+
+    rewriter.replaceOp(toElementsOp, replacements);
+    return success();
+  }
+};
```
Replicating the whole source element list only works when the broadcast duplicates entire blocks (e.g., only extra leading dims). When a 1-sized source dim sits inside the aligned shape, the destination order interleaves repetitions, so you must compute the per-index mapping.

An Example where naive replication is wrong:

Source: `vector<2x1x3xf32> → flatten indices: [0,1,2,3,4,5]`

Dest: `vector<2x4x3xf32> (middle dim broadcast from 1 to 4)`

Expected element order in dest (mapping each (i,j,k) to source (i,0,k)):

`[0,1,2, 0,1,2, 0,1,2, 0,1,2, 3,4,5, 3,4,5, 3,4,5, 3,4,5]`

Naively replicating the entire source list 4 times gives:

`[0,1,2,3,4,5, 0,1,2,3,4,5, 0,1,2,3,4,5, 0,1,2,3,4,5]`

So we need the decode-and-map step to produce the correct interleaving for all broadcast shapes.

keshavvinayak01 committed:
Addressed comments:
1. Added better docs for the folder with an example
2. Removed isScalable check, not required for toElementsOp
3. Used free create method for new toElementsOp

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

dcaballe reviewed (commented):
Thanks! It's an interesting canonicalization! A few comments

keshavvinayak01 committed:
Addressed review comments:
1. Added better doc for inner broadcast case
2. Added lit test for inner broadcast dim.

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

kuhar commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -2410,6 +2440,94 @@ ToElementsOp::inferReturnTypes(MLIRContext *ctx, std::optional<Location> loc,
+
+    SmallVector<int64_t> dstIdx(dstShape.size());
+    for (int64_t lin = 0; lin < dstCount; ++lin) {
+      int64_t temp = lin;
+      for (int64_t i = dstShape.size() - 1; i >= 0; --i) {
+        int64_t dim = dstShape[i];
+        dstIdx[i] = temp % dim;
+        temp /= dim;
+      }
+      int64_t srcLin = 0;
+      for (int64_t k = 0; k < srcRank; ++k)
+        srcLin = srcLin * srcShape[k] +
+                 ((srcShape[k] == 1) ? 0 : dstIdx[dstRank - srcRank + k]);
+
```
Are there no existing helpers we could use for this? We might have similar code for insert/extract folders, but I don't know if we can reuse it here.

keshavvinayak01 replied:
Phew! Had to sit and rewrite the entire computation using some existing utils. Added comments for clarity too.

keshavvinayak01 committed:
Rewrote computations with existing utils from IndexingUtils

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Merge branch 'main' into users/keshavvinayak01/toelements-broadcast-fold

kuhar reviewed (approved):
LGTM % nit + please remove the `#Describtion` header from the PR description / commit message -- it's obvious that it's a description.

keshavvinayak01 committed:
Update mlir/lib/Dialect/Vector/IR/VectorOps.cpp

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>

keshavvinayak01 committed:
Update mlir/lib/Dialect/Vector/IR/VectorOps.cpp

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>

DavidSpickett commented:
This has caused test failures on AArch64: https://lab.llvm.org/buildbot/#/builders/143/builds/11524

One example:
```
******************** TEST 'MLIR :: Integration/Dialect/Tensor/dim-runtime-verification.mlir' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
/home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/bin/mlir-opt /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/llvm/mlir/test/Integration/Dialect/Tensor/dim-runtime-verification.mlir -generate-runtime-verification      -one-shot-bufferize="bufferize-function-boundaries"      -buffer-deallocation-pipeline=private-function-dynamic-ownership      -test-cf-assert      -convert-to-llvm |  /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/bin/mlir-runner -e main -entry-point-result=void      -shared-libs=/home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/tools/mlir/test/Integration/Dialect/Tensor/Output/dim-runtime-verification.mlir.tmplir_runner_utils 2>&1 |  /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/bin/FileCheck /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/llvm/mlir/test/Integration/Dialect/Tensor/dim-runtime-verification.mlir
# executed command: /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/bin/mlir-opt /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/llvm/mlir/test/Integration/Dialect/Tensor/dim-runtime-verification.mlir -generate-runtime-verification -one-shot-bufferize=bufferize-function-boundaries -buffer-deallocation-pipeline=private-function-dynamic-ownership -test-cf-assert -convert-to-llvm
# executed command: /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/bin/mlir-runner -e main -entry-point-result=void -shared-libs=/home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/tools/mlir/test/Integration/Dialect/Tensor/Output/dim-runtime-verification.mlir.tmplir_runner_utils
# executed command: /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/bin/FileCheck /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/llvm/mlir/test/Integration/Dialect/Tensor/dim-runtime-verification.mlir
# .---command stderr------------
# | /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/llvm/mlir/test/Integration/Dialect/Tensor/dim-runtime-verification.mlir:25:17: error: CHECK-NEXT: expected string not found in input
# |  // CHECK-NEXT: "tensor.dim"(%{{.*}}, %{{.*}}) : (tensor<1xf32>, index) -> index
# |                 ^
# | <stdin>:3:38: note: scanning from here
# | ERROR: Runtime op verification failed
# |                                      ^
# | <stdin>:4:8: note: possible intended match here
# | %dim = tensor.dim %0, %c4 : tensor<1xf32>
# |        ^
# | 
# | Input file: <stdin>
# | Check file: /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/llvm/mlir/test/Integration/Dialect/Tensor/dim-runtime-verification.mlir
# | 
# | -dump-input=help explains the following input dump.
# | 
# | Input was:
# | <<<<<<
# |            1: Failed to create MemoryBuffer for: /home/tcwg-buildbot/worker/clang-aarch64-sve-vls/stage1/tools/mlir/test/Integration/Dialect/Tensor/Output/dim-runtime-verification.mlir.tmplir_runner_utils 
# |            2: Error: No such file or directory 
# |            3: ERROR: Runtime op verification failed 
# | next:25'0                                          X error: no match found
# |            4: %dim = tensor.dim %0, %c4 : tensor<1xf32> 
# | next:25'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | next:25'1            ?                                   possible intended match
# |            5: ^ 
# | next:25'0     ~~
# |            6: Location: loc("/home/tcwg-buildbot/worker/clang-aarch64-sve-vls/llvm/mlir/test/Integration/Dialect/Tensor/dim-runtime-verification.mlir":28:10) 
# | next:25'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | >>>>>>
# `-----------------------------
# error: command failed with exit status: 1

--

********************
```
The host does have SVE vector extensions, and if this is running code, it's running natively.

Also please set your email settings (https://llvm.org/docs/DeveloperPolicy.html#email-addresses) for future contributions so that you are notified.

(notifications are only posted on PRs if that PR is the only change included in a build)
